[Woo POS] Coupons: Products/Coupons Header

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Creating a header view for switching between Products and Coupons

1. Updated `POSPageHeaderView` to support multiple titles with action
2. Created `POSPageHeaderActionButton` for displaying a round action button
3. Updated `ItemListView` to display a new header if the feature flag is enabled

Info button is deliberately removed based on the designs and discussions during the design phase.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Enable `enableCouponsInPointOfSale` feature flag

1. Open POS
2. Compare the header with the designs FnDpCHVQ3Zpefad04ptbd4-fi-297_13359
3. Confirm switching between Products and Coupons continue to work
5. Confirm adding coupon action "+" continues to work

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on iPad Air 11 Inch iOS 18.4 Simulator

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Header View

https://github.com/user-attachments/assets/0ca7bcb7-38d3-464a-9f9e-5e36317a1da6

### Hidden Simple Products Banner

https://github.com/user-attachments/assets/98d48a83-b0f7-4e99-90b9-95f43bec9a88

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Add multiple title support to POSPageHeaderView

- Accept multiple items with titles, subtitles, and action
- Change selected item index on action

staskus committed:
Configure accessibility for POSPageHeaderView title

staskus committed:
Add Products and Coupons to PageHeader in ItemListView

staskus committed:
Create POSPageHeaderActionButton for showing quick actions

staskus committed:
Add coupon creation action button to the header

staskus committed:
Hide simple products banner from coupons

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/242">22.2</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -157,33 +163,28 @@ private extension ItemListView {
-                .foregroundStyle(Color.posOnSurface)
-            }
+    var headerViewItems: [POSPageHeaderItem] {
+        var items = [
+            POSPageHeaderItem(
+                title: Localization.productsTitle,
+                action: {
+                    displayItemListType(.products(search: searchTerm.isNotEmpty))
+                }
+            )
+        ]
+
+        if shouldShowCoupons {
+            items.append(
+                POSPageHeaderItem(
+                    title: Localization.couponsTitle,
+                    action: {
+                        displayItemListType(.coupons)
+                    }
+                )
+            )
```
Nice and clean at the callsite

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSPageHeaderActionButton.swift:
```diff
@@ -0,0 +1,20 @@
+                        .font(.posButtonSymbolSmall)
+                        .foregroundColor(.posOnSurface)
+                }
+                .frame(width: POSHeaderLayoutConstants.minHeight, height: POSHeaderLayoutConstants.minHeight)
```
This starts to look a little odd at larger dynamic type sizes. Perhaps use a `@ScaledMetric` inited with the min height, rather than the raw value? You can use the `min` of that and the constant, to avoid the button getting too small.

staskus replied:
I tried playing with ScaledMetric, likely min and max values are required since it also grows a bit too gigantic. I'll try to combine them.

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSPageHeaderView.swift:
```diff
@@ -12,30 +12,54 @@ struct POSPageHeaderBackButtonConfiguration {
 
     private var hStackAlignment: VerticalAlignment {
-        subtitle == nil ? .center: .firstTextBaseline
+        items.first?.subtitle == nil ? .center: .firstTextBaseline
```
Is the subtitle going to be used for something soon, or just future proofing? I can't see where it is at the moment.

staskus replied:
This component was part of the Design System work, so I kept the behavior for the multi-title scenario as well. The subtitle is used in the cash view to show the order total. 

![image](https://github.com/user-attachments/assets/ea88c27e-636d-4e69-9f4d-99f7191d1db6)

joshheald reviewed (approved):
Works as expected and matches designs. 

It gets pretty funky when the search field is shown at any dynamic type size; you get an animation effect on the titles changing size, as they're smaller when they're constrained by the search field, but then grow when you switch to coupons. That's not a blocker because you won't be able to switch tabs in search mode.

You can see a subtler version of the same issue when at very large dynamic sizes and constrained width, but it's acceptable for the extreme text size used. It's caused by the `+` button only being shown on Coupons.


https://github.com/user-attachments/assets/c998393e-c4e4-4a44-99ee-951483b207de

staskus commented:
> You can see a subtler version of the same issue when at very large dynamic sizes and constrained width, but it's acceptable for the extreme text size used. It's caused by the + button only being shown on Coupons.

@joshheald, thanks, good observations.

We could reserve some space for action buttons by keeping them invisible rather than removing them, then the amount of space for headers wouldn't change. 

Either way, this header is not great for extreme dynamic type sizes. Adding a third header would be enough to break it or it would require a different layout.

 If I look at native Apple apps, they do limit the dynamic type growth in quite a few places. Tabs at the bottom and top grow quite little when even the default font is much smaller than ours.

We need to do user testing, but I think large elements should grow relatively less, given they are already large.

staskus committed:
Merge branch 'trunk' into woomob-254-woo-pos-coupons-productscoupons-tabs-switch-header-designer

staskus committed:
Reserve space for coupon add action button

Hide the button with opacity rather than remove from hierarchy so the space would always be reserved for this button. It keeps the header titles size consistent when switching between the products and coupons while having large dynamic type sizes.

staskus committed:
Make POSPageHeaderActionButton scalable

Constraint between original and 1.5x times the size for the best UX

joshheald commented:
> If I look at native Apple apps, they do limit the dynamic type growth in quite a few places. Tabs at the bottom and top grow quite little when even the default font is much smaller than ours.
> 
> We need to do user testing, but I think large elements should grow relatively less, given they are already large.

I think that titles/tabs which are part of a small, consistent are also less likely to be read in detail. Error message text, for example, is much more important to fully scale as it'll happen rarely and the user needs to know what's going on.
